app/src/ui/clone-repository/clone-repository.tsx

```diff
@@ -0,0 +1,376 @@
+import * as Path from 'path'
+import * as React from 'react'
+import { remote } from 'electron'
+import { Button } from '../lib/button'
+import { ButtonGroup } from '../lib/button-group'
+import { Dispatcher } from '../../lib/dispatcher'
+import { getDefaultDir, setDefaultDir } from '../lib/default-dir'
+import { Account } from '../../models/account'
+import {
+  IRepositoryIdentifier,
+  parseRepositoryIdentifier,
+} from '../../lib/remote-parsing'
+import { findAccountForRemoteURL } from '../../lib/find-account'
+import { API } from '../../lib/api'
+import { Dialog, DialogError, DialogFooter, DialogContent } from '../dialog'
+import { TabBar } from '../tab-bar'
+import { CloneRepositoryTab } from '../../models/clone-repository-tab'
+import { CloneGenericRepository } from './clone-generic-repository'
+import { CloneGithubRepository } from './clone-github-repository'
+import { pathExists } from '../../lib/file-system'
+import { assertNever } from '../../lib/fatal-error'
+import { CallToAction } from '../lib/call-to-action'
+
+/** The name for the error when the destination already exists. */
+const DestinationExistsErrorName = 'DestinationExistsError'
+
+interface ICloneRepositoryProps {
+  readonly dispatcher: Dispatcher
+  readonly onDismissed: () => void
+
+  /** The logged in accounts. */
+  readonly dotComAccount: Account | null
+
+  /** The logged in Enterprise account. */
+  readonly enterpriseAccount: Account | null
+
+  /** The initial URL or `owner/name` shortcut to use. */
+  readonly initialURL: string | null
+
+  /** The initial tab to open on load */
+  readonly initialSelectedTab?: CloneRepositoryTab
+}
+
+interface ICloneRepositoryState {
+  /** The user-entered URL or `owner/name` shortcut. */
+  readonly url: string
+
+  /** The local path to clone to. */
+  readonly path: string
+
+  /** Are we currently trying to load the entered repository? */
+  readonly loading: boolean
+
+  /** The current error if one occurred. */
+  readonly error: Error | null
+
+  /**
+   * The repository identifier that was last parsed from the user-entered URL.
+   */
+  readonly lastParsedIdentifier: IRepositoryIdentifier | null
+
+  /**
+   * The default tab to open on load
+   */
+  readonly selectedTab: CloneRepositoryTab
+}
+
+/** The component for cloning a repository. */
+export class CloneRepository extends React.Component<
+  ICloneRepositoryProps,
+  ICloneRepositoryState
+> {
+  public constructor(props: ICloneRepositoryProps) {
+    super(props)
+
+    this.state = {
+      url: this.props.initialURL || '',
+      path: getDefaultDir(),
+      loading: false,
+      error: null,
+      lastParsedIdentifier: null,
+      selectedTab: this.props.initialSelectedTab || CloneRepositoryTab.DotCom,
+    }
+  }
+
+  public componentWillReceiveProps(nextProps: ICloneRepositoryProps) {
+    if (
+      this.state.selectedTab !== nextProps.initialSelectedTab &&
+      nextProps.initialSelectedTab
```

Comment:
I'll be honest, I struggled when I was thinking through this, but I don't _think_ so... consider:

1. User clicks "Open in Desktop" so the `initialSelectedTab` is set to the URL tab.
2. User switches to a different tab.
3. User clicks "Open in Desktop" again. `initialSelectedTab` would still be the URL tab, so it wouldn't switch.

Reply:
Hmm. I see your point but the current behavior does create some potential weirdness since it'll always trust the props and change the tab whenever `componentWillReceiveProps` changes (which, since this isn't a Pure component will be semi-often). Check this behavior out when switching app focus in order to trigger an app-store update:

![weirdness](https://user-images.githubusercontent.com/634063/30213913-e2077ece-94aa-11e7-9944-3014a9f020e9.gif)

Reply:
Bleh, right. I suppose the Right Way to fix this would be to push the select tab up into app state. But I'm not sure the case I pointed out matters enough to merit that. Thoughts?

Reply:
Right, either push it up to AppState or convert it to a sentinel value so that we can detect when it changes. Like `initialSelectedTab: { tabIndex: number }`. Not pretty but it'd probably work